### ISSUE TITLE
[C#] Implement new Concept Exercise: performance

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1018]
This issue describes how to implement the `performance` concept exercise for the C# track.

## Goal

The goal of this exercise is to teach the student the Concept of Performance in C#.

## Learning objectives

- Know what the garbage collector is and how it influences performance.
- Know about the memory differences between the stack and heap.
- Know how reference and value types relate to the memory and the garbage collector.
- Know how to define ref performance.
- Know how to use ref returns.
- Know how to use ref locals.
- Know about the `Span<T>` and `Memory<T>` types and their readonly variants.
- Know what `stackalloc` is and how to use it.

## Out of scope

- Memory pooling.
- Memory locality.
- `unsafe` code.
- `fixed` statements.
- Defensive copies and `readonly` semantics.
- Performance analysis/profiling.
- Discussion of performance tools (referring to them is fine).

## Concepts

This Concepts Exercise's Concepts are:

- `performance-basic`: know what the garbage collector is and how it influences performance; know about the memory differences between the stack and heap; know how reference and value types relate to the memory and the garbage collector; know how to define ref performance; know how to use ref returns; know how to use ref locals; know about the `Span<T>` and `Memory<T>` types and their readonly variants; know what `stackalloc` is and how to use it.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `references-basic`: know about value and reference types and passing parameters by reference.
- `classes-basic`: know how to define and work with classes.
- `performance-basic`: know how to define and work with performance.

### Hints

- [`Span<T>`][span]: describes the `Span<T>` class.
- [`ReadOnlySpan<T>`][readonly-span]: describes the `ReadOnlySpan<T>` class.
- [`Memory<T>`][memory]: describes the `Memory<T>` class.
- [`ReadOnlyMemory<T>`][readonly-memory]: describes the `ReadOnlyMemory<T>` class.
- [ref returns and ref locals][ref-returns-and-locals]: shows how to use ref returns and ref locals.
- [stackalloc][stackalloc]: explains how the `stackalloc` operator works.

### After

- [`Span<T>`][span]: describes the `Span<T>` class.
- [`ReadOnlySpan<T>`][readonly-span]: describes the `ReadOnlySpan<T>` class.
- [`Memory<T>`][memory]: describes the `Memory<T>` class.
- [`ReadOnlyMemory<T>`][readonly-memory]: describes the `ReadOnlyMemory<T>` class.
- [ref returns and ref locals][ref-returns-and-locals]: shows how to use ref returns and ref locals.
- [stackalloc][stackalloc]: explains how the `stackalloc` operator works.
- [`Span<T>` and `Memory<T>` guidelines][span-and-memory]: guidelines on how to use the `Span<T>` and `Memory<T>` types.
- [Value types vs reference types][value-types-vs-reference-types]: deep dive into the implementation and performance aspects of value and reference types.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise could benefit from having an [analyzer][analyzer] that can comment on:

- Not using `stackalloc`.

## Implementing

If you'd like to work on implementing this exercise, the first step is to let us know through a comment on this issue, to prevent multiple people from working on the same exercise. If you have any questions while implementing the exercise, please also post them as comments in this issue.

Implementing the exercise means creating the following files:

<pre>
languages
└── csharp
    └── exercises
        └── concept
            └── performance
                ├── .docs
                |   ├── after.md
                |   ├── hints.md
                |   ├── instructions.md
                |   └── introduction.md
                ├── .meta
                |   ├── config.json
                |   ├── design.md
                |   └── Example.cs
                ├── Performance.csproj
                ├── Performance.cs
                └── PerformanceTests.cs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should be explicit about what the student should learn from the exercise, and provide a short, concise introduction to the concept(s). The aim is to give the student just enough context to figure things out themselves and solve the exercise, as research has shown that self-discovery is the most effective learning experience. Mentioning technical terms that the student can Google if they so want, is preferable over including any code samples or an extensive description. For example we might describe a string as a "Sequence of Unicode characters" or a "series of bytes" or "an object". Unless the student needs to understand the details of what those mean to be able to solve the exercise we should not give more info in this introduction - instead allowing the student to Google, ignore, or map their existing knowledge.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

The above four files are also all described in the [concept exercises document][concept-exercises].

## Step 5: update languages/csharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "performance",
        "uuid": "2ab247a9-bda0-4486-a097-52d4f39eed13",
        "concepts": ["performance-basic"],
        "prerequisites": [
          "references-basic",
          "classes-basic",
          "structs-basic"
        ]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the C# track:

- `Performance.csproj`: the C# project file.
- `PerformanceTests.cs`: the test suite.
- `Performance.cs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `.meta/Example.cs`: an example implementation that passes all the tests.

Check out [an existing exercise][exercise-example] to see what these files should look like.

## Step 7: update the general concept document

Add the exercise to the concept's shared document's `## Implementations` section ([example][reference-example]).

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][design-example]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][config.json-example]).

### Inspiration

When implementing this exericse, it can be very useful to look at [already implemented C# exercises][implemented-exercises].

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/performance.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[value-types-vs-reference-types]: https://adamsitnik.com/Value-Types-vs-Reference-Types/
[span]: https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netcore-3.1
[readonly-span]: https://docs.microsoft.com/en-us/dotnet/api/system.readonlyspan-1?view=netcore-3.1
[memory]: https://docs.microsoft.com/en-us/dotnet/api/system.memory-1?view=netcore-3.1
[readonly-memory]: https://docs.microsoft.com/en-us/dotnet/api/system.readonlymemory-1?view=netcore-3.1
[span-and-memory]: https://docs.microsoft.com/en-us/dotnet/standard/memory-and-spans/memory-t-usage-guidelines
[ref-returns-and-locals]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/ref-returns
[stackalloc]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/stackalloc

**Comment by mikedamay** [comment]
Should we consider closing this issue and combining the concept with `struct`s?  Exposure to the subtleties of heap vs. stack, large object heap, garbage collection might fit more easily in `struct`s' _after.md_.

**Comment by ErikSchierboom** [comment]
This is a tough call I feel. The reason why I had it as a separate exercise is because I think there is quite a bit of information here. It might be that the `after.md` of the `structs` exercise could be a bit dense. Maybe we could leave it open for now but still explore if we could add this information in the `structs` exercise's `after.md` file.

**Comment by mikedamay** [comment]
Sounds good.  Part of my reasoning was the difficulties of setting puzzles.   It's also related to a point I was trying to make under Exercism's Ambition [here](https://github.com/exercism/v3/issues/690#issuecomment-641066362).

**Comment by ErikSchierboom** [comment]
Yes, I read that! It is an excellent point.

**Comment by mikedamay** [comment]
- Contrast C# developers' attitude to out-of-memory conditions with those of lower level languages like C.
